{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e9c226ba-6272-4758-a166-801fa7d8eb7a",
   "metadata": {},
   "source": [
    "#### This section covers the process of getting the 500bp upstream and downstream sequences from the TSS of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "71cb8bb1-7980-4c60-bc9e-893b801997ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"C:/Users/muell/OneDrive/Documents/ML_FinalProject/final_gene_expression_data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "6b827b25-a2bb-475c-a0ab-07708a604050",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a mapping of cleaned IDs to original IDs\n",
    "df['cleaned_gene_ids'] = df['Name'].str.split('.').str[0]\n",
    "id_mapping = df[['cleaned_gene_ids', 'Name']].drop_duplicates()\n",
    "\n",
    "# Save the mapping\n",
    "id_mapping.to_csv(\"id_mapping.csv\", index=False)\n",
    "\n",
    "# Save the cleaned gene IDs\n",
    "id_mapping['cleaned_gene_ids'].to_csv(\"unique_gene_ids.txt\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "124fa5ae-d737-4b95-8bd0-a97b81762389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22394, 2)"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapping.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "a5c00765-368f-40e6-b800-75359ac39b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# GTF file details\n",
    "gtf_file = \"Homo_sapiens.GRCh38.113.gtf\"\n",
    "gtf_cols = [\"chromosome\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "\n",
    "# Specify column data types\n",
    "gtf_dtypes = {\n",
    "    \"chromosome\": \"string\",\n",
    "    \"source\": \"string\",\n",
    "    \"feature\": \"string\",\n",
    "    \"start\": \"Int64\", \n",
    "    \"end\": \"Int64\",\n",
    "    \"score\": \"string\",\n",
    "    \"strand\": \"string\",\n",
    "    \"frame\": \"string\",\n",
    "    \"attribute\": \"string\",\n",
    "}\n",
    "\n",
    "# Load the GTF file with specified dtypes\n",
    "gtf = pd.read_csv(gtf_file, sep=\"\\t\", comment=\"#\", names=gtf_cols, dtype=gtf_dtypes, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "77adeed6-ac25-409d-a76c-6b44ca0b603b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    chromosome          source feature     start       end score strand frame  \\\n",
      "0            1  ensembl_havana    gene   3069168   3438621     .      +     .   \n",
      "220          1          havana    gene   5301928   5307394     .      -     .   \n",
      "226          1  ensembl_havana    gene   2403964   2413797     .      -     .   \n",
      "346          1          havana    gene   5492978   5494674     .      +     .   \n",
      "349          1          havana    gene  10054445  10054781     .      -     .   \n",
      "\n",
      "                                             attribute          gene_id  \n",
      "0    gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  ENSG00000142611  \n",
      "220  gene_id \"ENSG00000284616\"; gene_version \"1\"; g...  ENSG00000284616  \n",
      "226  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  ENSG00000157911  \n",
      "346  gene_id \"ENSG00000260972\"; gene_version \"1\"; g...  ENSG00000260972  \n",
      "349  gene_id \"ENSG00000224340\"; gene_version \"1\"; g...  ENSG00000224340  \n"
     ]
    }
   ],
   "source": [
    "# Filter for \"gene\" features only\n",
    "genes = gtf[gtf[\"feature\"] == \"gene\"].copy()\n",
    "\n",
    "# Define a function for safe regex extraction\n",
    "def extract_gene_id(attribute):\n",
    "    match = re.search(r'gene_id \"([^\"]+)\"', attribute)\n",
    "    return match.group(1) if match else None\n",
    "\n",
    "# Extract the gene_id from the attribute column\n",
    "genes[\"gene_id\"] = genes[\"attribute\"].apply(extract_gene_id)\n",
    "\n",
    "# Verify the resulting DataFrame\n",
    "print(genes.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "f86664f8-2766-44dd-9fb6-f6ca2e3b421a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the unique gene IDs\n",
    "unique_gene_ids = pd.read_csv(\"unique_gene_ids.txt\", header=None, names=[\"gene_id\"])\n",
    "\n",
    "# Filter the GTF genes to include only those in the unique gene list\n",
    "filtered_genes = genes[genes[\"gene_id\"].isin(unique_gene_ids[\"gene_id\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "ad2ab268-8f25-48ae-a914-c67366d00821",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss_start</th>\n",
       "      <th>tss_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>2413297</td>\n",
       "      <td>2414297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "      <td>ENSG00000142655</td>\n",
       "      <td>10471788</td>\n",
       "      <td>10472788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2425980</td>\n",
       "      <td>2505532</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000149527\"; gene_version \"18\"; ...</td>\n",
       "      <td>ENSG00000149527</td>\n",
       "      <td>2425480</td>\n",
       "      <td>2426480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>676</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9292894</td>\n",
       "      <td>9369532</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000171621\"; gene_version \"14\"; ...</td>\n",
       "      <td>ENSG00000171621</td>\n",
       "      <td>9292394</td>\n",
       "      <td>9293394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2492300</td>\n",
       "      <td>2493258</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000224387\"; gene_version \"1\"; g...</td>\n",
       "      <td>ENSG00000224387</td>\n",
       "      <td>2492758</td>\n",
       "      <td>2493758</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome          source feature     start       end score strand frame  \\\n",
       "226          1  ensembl_havana    gene   2403964   2413797     .      -     .   \n",
       "355          1  ensembl_havana    gene  10472288  10630758     .      +     .   \n",
       "444          1  ensembl_havana    gene   2425980   2505532     .      +     .   \n",
       "676          1  ensembl_havana    gene   9292894   9369532     .      +     .   \n",
       "721          1          havana    gene   2492300   2493258     .      -     .   \n",
       "\n",
       "                                             attribute          gene_id  \\\n",
       "226  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  ENSG00000157911   \n",
       "355  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...  ENSG00000142655   \n",
       "444  gene_id \"ENSG00000149527\"; gene_version \"18\"; ...  ENSG00000149527   \n",
       "676  gene_id \"ENSG00000171621\"; gene_version \"14\"; ...  ENSG00000171621   \n",
       "721  gene_id \"ENSG00000224387\"; gene_version \"1\"; g...  ENSG00000224387   \n",
       "\n",
       "     tss_start   tss_end  \n",
       "226    2413297   2414297  \n",
       "355   10471788  10472788  \n",
       "444    2425480   2426480  \n",
       "676    9292394   9293394  \n",
       "721    2492758   2493758  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "57cdb2bd-bba3-4abc-919c-4a58d75b356a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to a new GTF-like file\n",
    "filtered_genes.to_csv(\"filtered_genes.gtf\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "b027a478-64b6-4ade-abd0-977fa9a084ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the required columns for BED format\n",
    "bed_df = filtered_genes[[\"chromosome\", \"tss_start\", \"tss_end\", \"gene_id\", \"score\", \"strand\"]]\n",
    "\n",
    "# Save to a BED file\n",
    "bed_df.to_csv(\"gtex_transcripts.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "1ac9c89e-d0b1-408d-bf02-e77194170c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load gene expression data\n",
    "expression_df = pd.read_csv(\"final_gene_expression_data.csv\")\n",
    "\n",
    "# Load the mapping file (gene_id to transcript_id)\n",
    "mapping_df = pd.read_csv(\"filtered_mapping.txt\", sep=\"\\t\", names=[\"gene_id\", \"transcript_id\"])\n",
    "\n",
    "# Load the sequences (transcript_id and sequence)\n",
    "sequences_df = pd.read_csv(\"gtex_sequences.csv\", header=None, names=[\"transcript_id\", \"sequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "9d123f02-a2cf-42a7-bc10-0881cfed50c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>PC9</th>\n",
       "      <th>PC10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>-0.123846</td>\n",
       "      <td>-0.013572</td>\n",
       "      <td>-0.002610</td>\n",
       "      <td>-0.008449</td>\n",
       "      <td>-0.000559</td>\n",
       "      <td>-0.008659</td>\n",
       "      <td>-0.001629</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.002948</td>\n",
       "      <td>0.000726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>-0.127444</td>\n",
       "      <td>-0.014391</td>\n",
       "      <td>-0.002761</td>\n",
       "      <td>-0.008784</td>\n",
       "      <td>-0.001466</td>\n",
       "      <td>-0.009513</td>\n",
       "      <td>-0.002162</td>\n",
       "      <td>0.001299</td>\n",
       "      <td>0.003319</td>\n",
       "      <td>0.000545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000268903.1</td>\n",
       "      <td>ENSG00000268903</td>\n",
       "      <td>0.073242</td>\n",
       "      <td>-0.012280</td>\n",
       "      <td>-0.001378</td>\n",
       "      <td>0.065349</td>\n",
       "      <td>0.011552</td>\n",
       "      <td>0.034967</td>\n",
       "      <td>-0.050337</td>\n",
       "      <td>-0.033125</td>\n",
       "      <td>-0.027376</td>\n",
       "      <td>0.001724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000269981.1</td>\n",
       "      <td>ENSG00000269981</td>\n",
       "      <td>0.197083</td>\n",
       "      <td>-0.010883</td>\n",
       "      <td>0.000580</td>\n",
       "      <td>0.087887</td>\n",
       "      <td>-0.012004</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>-0.053531</td>\n",
       "      <td>-0.069170</td>\n",
       "      <td>-0.050675</td>\n",
       "      <td>-0.023379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000279928.2</td>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>-0.128000</td>\n",
       "      <td>-0.014400</td>\n",
       "      <td>-0.002366</td>\n",
       "      <td>-0.009148</td>\n",
       "      <td>-0.001414</td>\n",
       "      <td>-0.009585</td>\n",
       "      <td>-0.001641</td>\n",
       "      <td>0.001489</td>\n",
       "      <td>0.003598</td>\n",
       "      <td>0.000606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Name      Description       PC1       PC2       PC3       PC4  \\\n",
       "0  ENSG00000227232.5           WASH7P -0.123846 -0.013572 -0.002610 -0.008449   \n",
       "1  ENSG00000233750.3           CICP27 -0.127444 -0.014391 -0.002761 -0.008784   \n",
       "2  ENSG00000268903.1  ENSG00000268903  0.073242 -0.012280 -0.001378  0.065349   \n",
       "3  ENSG00000269981.1  ENSG00000269981  0.197083 -0.010883  0.000580  0.087887   \n",
       "4  ENSG00000279928.2         DDX11L17 -0.128000 -0.014400 -0.002366 -0.009148   \n",
       "\n",
       "        PC5       PC6       PC7       PC8       PC9      PC10  \n",
       "0 -0.000559 -0.008659 -0.001629  0.001415  0.002948  0.000726  \n",
       "1 -0.001466 -0.009513 -0.002162  0.001299  0.003319  0.000545  \n",
       "2  0.011552  0.034967 -0.050337 -0.033125 -0.027376  0.001724  \n",
       "3 -0.012004  0.052083 -0.053531 -0.069170 -0.050675 -0.023379  \n",
       "4 -0.001414 -0.009585 -0.001641  0.001489  0.003598  0.000606  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "c5b1426a-1c8e-4443-975b-c85971abaf3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>ENST00000288774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>ENST00000447513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>ENST00000650293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>ENST00000507596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>ENST00000510434</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id\n",
       "0  ENSG00000157911  ENST00000288774\n",
       "1  ENSG00000157911  ENST00000447513\n",
       "2  ENSG00000157911  ENST00000650293\n",
       "3  ENSG00000157911  ENST00000507596\n",
       "4  ENSG00000157911  ENST00000510434"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "d0dd9c38-73ac-4feb-90e1-1ab00d305a54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000157911::1:2413297-2414297(-)</td>\n",
       "      <td>GGTGCAGATGAGAAGAGTGGGGAAAGCAGTCCTGAGCCAGGACAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000142655::1:10471788-10472788(+)</td>\n",
       "      <td>AGTGCAAGGCCAAGCAGCTGGTAAAGGTATGGGGCCGGGCTCGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000149527::1:2425480-2426480(+)</td>\n",
       "      <td>GCAGGAGTGGGGTTGCTGGATCTGCTCTGCTGGTTTTTAGGTGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000171621::1:9292394-9293394(+)</td>\n",
       "      <td>AAACGCGAGCGGCCGAAACAGCGCTGCAGGAGGCTCCCTGGACCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000224387::1:2492758-2493758(-)</td>\n",
       "      <td>ACGCGTTGGAAATCAGGGCACGTCGCGTCCGCAGGCCCCATGCCAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             transcript_id  \\\n",
       "0    ENSG00000157911::1:2413297-2414297(-)   \n",
       "1  ENSG00000142655::1:10471788-10472788(+)   \n",
       "2    ENSG00000149527::1:2425480-2426480(+)   \n",
       "3    ENSG00000171621::1:9292394-9293394(+)   \n",
       "4    ENSG00000224387::1:2492758-2493758(-)   \n",
       "\n",
       "                                            sequence  \n",
       "0  GGTGCAGATGAGAAGAGTGGGGAAAGCAGTCCTGAGCCAGGACAGC...  \n",
       "1  AGTGCAAGGCCAAGCAGCTGGTAAAGGTATGGGGCCGGGCTCGGAA...  \n",
       "2  GCAGGAGTGGGGTTGCTGGATCTGCTCTGCTGGTTTTTAGGTGGCA...  \n",
       "3  AAACGCGAGCGGCCGAAACAGCGCTGCAGGAGGCTCCCTGGACCCC...  \n",
       "4  ACGCGTTGGAAATCAGGGCACGTCGCGTCCGCAGGCCCCATGCCAG...  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "85b3b56e-44c1-4ade-bd3c-737a66492d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_df['cleaned_gene_id'] = expression_df['Name'].str.split('.').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "db1ae929-3631-418f-9441-108e71f88824",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean transcript_id in sequences_df\n",
    "sequences_df['gene_id'] = sequences_df['transcript_id'].str.split('::').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "48caf3ae-a63f-4d27-98c5-9409a07e2367",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000157911::1:2413297-2414297(-)</td>\n",
       "      <td>GGTGCAGATGAGAAGAGTGGGGAAAGCAGTCCTGAGCCAGGACAGC...</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000142655::1:10471788-10472788(+)</td>\n",
       "      <td>AGTGCAAGGCCAAGCAGCTGGTAAAGGTATGGGGCCGGGCTCGGAA...</td>\n",
       "      <td>ENSG00000142655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000149527::1:2425480-2426480(+)</td>\n",
       "      <td>GCAGGAGTGGGGTTGCTGGATCTGCTCTGCTGGTTTTTAGGTGGCA...</td>\n",
       "      <td>ENSG00000149527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000171621::1:9292394-9293394(+)</td>\n",
       "      <td>AAACGCGAGCGGCCGAAACAGCGCTGCAGGAGGCTCCCTGGACCCC...</td>\n",
       "      <td>ENSG00000171621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000224387::1:2492758-2493758(-)</td>\n",
       "      <td>ACGCGTTGGAAATCAGGGCACGTCGCGTCCGCAGGCCCCATGCCAG...</td>\n",
       "      <td>ENSG00000224387</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             transcript_id  \\\n",
       "0    ENSG00000157911::1:2413297-2414297(-)   \n",
       "1  ENSG00000142655::1:10471788-10472788(+)   \n",
       "2    ENSG00000149527::1:2425480-2426480(+)   \n",
       "3    ENSG00000171621::1:9292394-9293394(+)   \n",
       "4    ENSG00000224387::1:2492758-2493758(-)   \n",
       "\n",
       "                                            sequence          gene_id  \n",
       "0  GGTGCAGATGAGAAGAGTGGGGAAAGCAGTCCTGAGCCAGGACAGC...  ENSG00000157911  \n",
       "1  AGTGCAAGGCCAAGCAGCTGGTAAAGGTATGGGGCCGGGCTCGGAA...  ENSG00000142655  \n",
       "2  GCAGGAGTGGGGTTGCTGGATCTGCTCTGCTGGTTTTTAGGTGGCA...  ENSG00000149527  \n",
       "3  AAACGCGAGCGGCCGAAACAGCGCTGCAGGAGGCTCCCTGGACCCC...  ENSG00000171621  \n",
       "4  ACGCGTTGGAAATCAGGGCACGTCGCGTCCGCAGGCCCCATGCCAG...  ENSG00000224387  "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "a8a20711-68eb-4eba-b614-a6b1e3e000a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge expression_df with mapping_df on gene_id\n",
    "expression_with_transcripts = pd.merge(expression_df, mapping_df, left_on=\"cleaned_gene_id\", right_on=\"gene_id\", how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "e9981bb8-fc5d-4f3f-a5c2-8c876f742a0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>PC9</th>\n",
       "      <th>PC10</th>\n",
       "      <th>cleaned_gene_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>-0.123846</td>\n",
       "      <td>-0.013572</td>\n",
       "      <td>-0.002610</td>\n",
       "      <td>-0.008449</td>\n",
       "      <td>-0.000559</td>\n",
       "      <td>-0.008659</td>\n",
       "      <td>-0.001629</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.002948</td>\n",
       "      <td>0.000726</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000488147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>-0.127444</td>\n",
       "      <td>-0.014391</td>\n",
       "      <td>-0.002761</td>\n",
       "      <td>-0.008784</td>\n",
       "      <td>-0.001466</td>\n",
       "      <td>-0.009513</td>\n",
       "      <td>-0.002162</td>\n",
       "      <td>0.001299</td>\n",
       "      <td>0.003319</td>\n",
       "      <td>0.000545</td>\n",
       "      <td>ENSG00000233750</td>\n",
       "      <td>ENSG00000233750</td>\n",
       "      <td>ENST00000442987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000268903.1</td>\n",
       "      <td>ENSG00000268903</td>\n",
       "      <td>0.073242</td>\n",
       "      <td>-0.012280</td>\n",
       "      <td>-0.001378</td>\n",
       "      <td>0.065349</td>\n",
       "      <td>0.011552</td>\n",
       "      <td>0.034967</td>\n",
       "      <td>-0.050337</td>\n",
       "      <td>-0.033125</td>\n",
       "      <td>-0.027376</td>\n",
       "      <td>0.001724</td>\n",
       "      <td>ENSG00000268903</td>\n",
       "      <td>ENSG00000268903</td>\n",
       "      <td>ENST00000494149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000269981.1</td>\n",
       "      <td>ENSG00000269981</td>\n",
       "      <td>0.197083</td>\n",
       "      <td>-0.010883</td>\n",
       "      <td>0.000580</td>\n",
       "      <td>0.087887</td>\n",
       "      <td>-0.012004</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>-0.053531</td>\n",
       "      <td>-0.069170</td>\n",
       "      <td>-0.050675</td>\n",
       "      <td>-0.023379</td>\n",
       "      <td>ENSG00000269981</td>\n",
       "      <td>ENSG00000269981</td>\n",
       "      <td>ENST00000595919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000279928.2</td>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>-0.128000</td>\n",
       "      <td>-0.014400</td>\n",
       "      <td>-0.002366</td>\n",
       "      <td>-0.009148</td>\n",
       "      <td>-0.001414</td>\n",
       "      <td>-0.009585</td>\n",
       "      <td>-0.001641</td>\n",
       "      <td>0.001489</td>\n",
       "      <td>0.003598</td>\n",
       "      <td>0.000606</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>ENST00000624431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Name      Description       PC1       PC2       PC3       PC4  \\\n",
       "0  ENSG00000227232.5           WASH7P -0.123846 -0.013572 -0.002610 -0.008449   \n",
       "1  ENSG00000233750.3           CICP27 -0.127444 -0.014391 -0.002761 -0.008784   \n",
       "2  ENSG00000268903.1  ENSG00000268903  0.073242 -0.012280 -0.001378  0.065349   \n",
       "3  ENSG00000269981.1  ENSG00000269981  0.197083 -0.010883  0.000580  0.087887   \n",
       "4  ENSG00000279928.2         DDX11L17 -0.128000 -0.014400 -0.002366 -0.009148   \n",
       "\n",
       "        PC5       PC6       PC7       PC8       PC9      PC10  \\\n",
       "0 -0.000559 -0.008659 -0.001629  0.001415  0.002948  0.000726   \n",
       "1 -0.001466 -0.009513 -0.002162  0.001299  0.003319  0.000545   \n",
       "2  0.011552  0.034967 -0.050337 -0.033125 -0.027376  0.001724   \n",
       "3 -0.012004  0.052083 -0.053531 -0.069170 -0.050675 -0.023379   \n",
       "4 -0.001414 -0.009585 -0.001641  0.001489  0.003598  0.000606   \n",
       "\n",
       "   cleaned_gene_id          gene_id    transcript_id  \n",
       "0  ENSG00000227232  ENSG00000227232  ENST00000488147  \n",
       "1  ENSG00000233750  ENSG00000233750  ENST00000442987  \n",
       "2  ENSG00000268903  ENSG00000268903  ENST00000494149  \n",
       "3  ENSG00000269981  ENSG00000269981  ENST00000595919  \n",
       "4  ENSG00000279928  ENSG00000279928  ENST00000624431  "
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_with_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "4d8e431a-5f15-4679-9df3-2dbd078a92a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_merged_df = pd.merge(expression_with_transcripts, sequences_df, on=\"gene_id\", how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "9df8eae9-7cdf-4d70-8c03-6cbe341ff9ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>PC9</th>\n",
       "      <th>PC10</th>\n",
       "      <th>cleaned_gene_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id_x</th>\n",
       "      <th>transcript_id_y</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>-0.123846</td>\n",
       "      <td>-0.013572</td>\n",
       "      <td>-0.002610</td>\n",
       "      <td>-0.008449</td>\n",
       "      <td>-0.000559</td>\n",
       "      <td>-0.008659</td>\n",
       "      <td>-0.001629</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.002948</td>\n",
       "      <td>0.000726</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>ENSG00000227232::1:24386-25386(-)</td>\n",
       "      <td>AACCCGTGTGTGCACATTTTTTGTGCTTTTCCAGAACTGTGTACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>-0.127444</td>\n",
       "      <td>-0.014391</td>\n",
       "      <td>-0.002761</td>\n",
       "      <td>-0.008784</td>\n",
       "      <td>-0.001466</td>\n",
       "      <td>-0.009513</td>\n",
       "      <td>-0.002162</td>\n",
       "      <td>0.001299</td>\n",
       "      <td>0.003319</td>\n",
       "      <td>0.000545</td>\n",
       "      <td>ENSG00000233750</td>\n",
       "      <td>ENSG00000233750</td>\n",
       "      <td>ENST00000442987</td>\n",
       "      <td>ENSG00000233750::1:130525-131525(+)</td>\n",
       "      <td>GTCAAAAGTTAAAGAAACTTTGCAGGTCAGGACAGAATCGAATAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000268903.1</td>\n",
       "      <td>ENSG00000268903</td>\n",
       "      <td>0.073242</td>\n",
       "      <td>-0.012280</td>\n",
       "      <td>-0.001378</td>\n",
       "      <td>0.065349</td>\n",
       "      <td>0.011552</td>\n",
       "      <td>0.034967</td>\n",
       "      <td>-0.050337</td>\n",
       "      <td>-0.033125</td>\n",
       "      <td>-0.027376</td>\n",
       "      <td>0.001724</td>\n",
       "      <td>ENSG00000268903</td>\n",
       "      <td>ENSG00000268903</td>\n",
       "      <td>ENST00000494149</td>\n",
       "      <td>ENSG00000268903::1:135395-136395(-)</td>\n",
       "      <td>TCACACCGGCCCCTCCCACGCTGAGAGAGGTCAGTGTGAGCCCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000269981.1</td>\n",
       "      <td>ENSG00000269981</td>\n",
       "      <td>0.197083</td>\n",
       "      <td>-0.010883</td>\n",
       "      <td>0.000580</td>\n",
       "      <td>0.087887</td>\n",
       "      <td>-0.012004</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>-0.053531</td>\n",
       "      <td>-0.069170</td>\n",
       "      <td>-0.050675</td>\n",
       "      <td>-0.023379</td>\n",
       "      <td>ENSG00000269981</td>\n",
       "      <td>ENSG00000269981</td>\n",
       "      <td>ENST00000595919</td>\n",
       "      <td>ENSG00000269981::1:137465-138465(-)</td>\n",
       "      <td>CAAATCAGGCTTTTGCCCAACTTCTGTCTACTGTCGGACTCTACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000279928.2</td>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>-0.128000</td>\n",
       "      <td>-0.014400</td>\n",
       "      <td>-0.002366</td>\n",
       "      <td>-0.009148</td>\n",
       "      <td>-0.001414</td>\n",
       "      <td>-0.009585</td>\n",
       "      <td>-0.001641</td>\n",
       "      <td>0.001489</td>\n",
       "      <td>0.003598</td>\n",
       "      <td>0.000606</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>ENSG00000279928::1:182196-183196(+)</td>\n",
       "      <td>CCCTGGAGATTCTTATTAGTGATTTGGGCTGGGGCCTGGCCATGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193942</th>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>1.288162</td>\n",
       "      <td>0.257889</td>\n",
       "      <td>-0.012438</td>\n",
       "      <td>0.160438</td>\n",
       "      <td>0.039930</td>\n",
       "      <td>0.555881</td>\n",
       "      <td>0.393585</td>\n",
       "      <td>-0.420538</td>\n",
       "      <td>0.035647</td>\n",
       "      <td>0.134114</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>ENSG00000198695::MT:14173-15173(-)</td>\n",
       "      <td>CCCCTCAGAATGATATTTGGCCTCACGGGAGGACATAGCCTATGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193943</th>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>-0.124354</td>\n",
       "      <td>-0.013892</td>\n",
       "      <td>-0.002712</td>\n",
       "      <td>-0.008277</td>\n",
       "      <td>-0.001226</td>\n",
       "      <td>-0.008314</td>\n",
       "      <td>-0.000863</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>0.003963</td>\n",
       "      <td>0.001268</td>\n",
       "      <td>ENSG00000210194</td>\n",
       "      <td>ENSG00000210194</td>\n",
       "      <td>ENST00000387459</td>\n",
       "      <td>ENSG00000210194::MT:14242-15242(-)</td>\n",
       "      <td>CTCAGATTCATTGAACTAGGTCTGTCCCAATGTATGGGATGGCGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193944</th>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>3.957295</td>\n",
       "      <td>2.872498</td>\n",
       "      <td>0.318762</td>\n",
       "      <td>0.160265</td>\n",
       "      <td>0.302238</td>\n",
       "      <td>1.740722</td>\n",
       "      <td>0.944125</td>\n",
       "      <td>-0.666716</td>\n",
       "      <td>0.313799</td>\n",
       "      <td>0.260474</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>ENST00000361789</td>\n",
       "      <td>ENSG00000198727::MT:14247-15247(+)</td>\n",
       "      <td>CGCACCAATAGGATCCTCCCGAATCAACCCTGACCCCTCTCCTTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193945</th>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>-0.128244</td>\n",
       "      <td>-0.014401</td>\n",
       "      <td>-0.002581</td>\n",
       "      <td>-0.009073</td>\n",
       "      <td>-0.001497</td>\n",
       "      <td>-0.009844</td>\n",
       "      <td>-0.001851</td>\n",
       "      <td>0.001706</td>\n",
       "      <td>0.003488</td>\n",
       "      <td>0.000751</td>\n",
       "      <td>ENSG00000210195</td>\n",
       "      <td>ENSG00000210195</td>\n",
       "      <td>ENST00000387460</td>\n",
       "      <td>ENSG00000210195::MT:15388-16388(+)</td>\n",
       "      <td>TCCGATAAAATCACCTTCCACCCTTACTACACAATCAAAGACGCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193946</th>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>-0.128048</td>\n",
       "      <td>-0.014273</td>\n",
       "      <td>-0.002491</td>\n",
       "      <td>-0.009159</td>\n",
       "      <td>-0.001423</td>\n",
       "      <td>-0.009515</td>\n",
       "      <td>-0.001758</td>\n",
       "      <td>0.001627</td>\n",
       "      <td>0.003683</td>\n",
       "      <td>0.000778</td>\n",
       "      <td>ENSG00000210196</td>\n",
       "      <td>ENSG00000210196</td>\n",
       "      <td>ENST00000387461</td>\n",
       "      <td>ENSG00000210196::MT:15523-16523(-)</td>\n",
       "      <td>TATGACCCTGAAGTAGGAACCAGATGTCGGATACAGTTCACTTTAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193947 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Name      Description       PC1       PC2       PC3  \\\n",
       "0       ENSG00000227232.5           WASH7P -0.123846 -0.013572 -0.002610   \n",
       "1       ENSG00000233750.3           CICP27 -0.127444 -0.014391 -0.002761   \n",
       "2       ENSG00000268903.1  ENSG00000268903  0.073242 -0.012280 -0.001378   \n",
       "3       ENSG00000269981.1  ENSG00000269981  0.197083 -0.010883  0.000580   \n",
       "4       ENSG00000279928.2         DDX11L17 -0.128000 -0.014400 -0.002366   \n",
       "...                   ...              ...       ...       ...       ...   \n",
       "193942  ENSG00000198695.2           MT-ND6  1.288162  0.257889 -0.012438   \n",
       "193943  ENSG00000210194.1            MT-TE -0.124354 -0.013892 -0.002712   \n",
       "193944  ENSG00000198727.2           MT-CYB  3.957295  2.872498  0.318762   \n",
       "193945  ENSG00000210195.2            MT-TT -0.128244 -0.014401 -0.002581   \n",
       "193946  ENSG00000210196.2            MT-TP -0.128048 -0.014273 -0.002491   \n",
       "\n",
       "             PC4       PC5       PC6       PC7       PC8       PC9      PC10  \\\n",
       "0      -0.008449 -0.000559 -0.008659 -0.001629  0.001415  0.002948  0.000726   \n",
       "1      -0.008784 -0.001466 -0.009513 -0.002162  0.001299  0.003319  0.000545   \n",
       "2       0.065349  0.011552  0.034967 -0.050337 -0.033125 -0.027376  0.001724   \n",
       "3       0.087887 -0.012004  0.052083 -0.053531 -0.069170 -0.050675 -0.023379   \n",
       "4      -0.009148 -0.001414 -0.009585 -0.001641  0.001489  0.003598  0.000606   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "193942  0.160438  0.039930  0.555881  0.393585 -0.420538  0.035647  0.134114   \n",
       "193943 -0.008277 -0.001226 -0.008314 -0.000863  0.000342  0.003963  0.001268   \n",
       "193944  0.160265  0.302238  1.740722  0.944125 -0.666716  0.313799  0.260474   \n",
       "193945 -0.009073 -0.001497 -0.009844 -0.001851  0.001706  0.003488  0.000751   \n",
       "193946 -0.009159 -0.001423 -0.009515 -0.001758  0.001627  0.003683  0.000778   \n",
       "\n",
       "        cleaned_gene_id          gene_id  transcript_id_x  \\\n",
       "0       ENSG00000227232  ENSG00000227232  ENST00000488147   \n",
       "1       ENSG00000233750  ENSG00000233750  ENST00000442987   \n",
       "2       ENSG00000268903  ENSG00000268903  ENST00000494149   \n",
       "3       ENSG00000269981  ENSG00000269981  ENST00000595919   \n",
       "4       ENSG00000279928  ENSG00000279928  ENST00000624431   \n",
       "...                 ...              ...              ...   \n",
       "193942  ENSG00000198695  ENSG00000198695  ENST00000361681   \n",
       "193943  ENSG00000210194  ENSG00000210194  ENST00000387459   \n",
       "193944  ENSG00000198727  ENSG00000198727  ENST00000361789   \n",
       "193945  ENSG00000210195  ENSG00000210195  ENST00000387460   \n",
       "193946  ENSG00000210196  ENSG00000210196  ENST00000387461   \n",
       "\n",
       "                            transcript_id_y  \\\n",
       "0         ENSG00000227232::1:24386-25386(-)   \n",
       "1       ENSG00000233750::1:130525-131525(+)   \n",
       "2       ENSG00000268903::1:135395-136395(-)   \n",
       "3       ENSG00000269981::1:137465-138465(-)   \n",
       "4       ENSG00000279928::1:182196-183196(+)   \n",
       "...                                     ...   \n",
       "193942   ENSG00000198695::MT:14173-15173(-)   \n",
       "193943   ENSG00000210194::MT:14242-15242(-)   \n",
       "193944   ENSG00000198727::MT:14247-15247(+)   \n",
       "193945   ENSG00000210195::MT:15388-16388(+)   \n",
       "193946   ENSG00000210196::MT:15523-16523(-)   \n",
       "\n",
       "                                                 sequence  \n",
       "0       AACCCGTGTGTGCACATTTTTTGTGCTTTTCCAGAACTGTGTACCA...  \n",
       "1       GTCAAAAGTTAAAGAAACTTTGCAGGTCAGGACAGAATCGAATAAT...  \n",
       "2       TCACACCGGCCCCTCCCACGCTGAGAGAGGTCAGTGTGAGCCCTTG...  \n",
       "3       CAAATCAGGCTTTTGCCCAACTTCTGTCTACTGTCGGACTCTACAG...  \n",
       "4       CCCTGGAGATTCTTATTAGTGATTTGGGCTGGGGCCTGGCCATGTG...  \n",
       "...                                                   ...  \n",
       "193942  CCCCTCAGAATGATATTTGGCCTCACGGGAGGACATAGCCTATGAA...  \n",
       "193943  CTCAGATTCATTGAACTAGGTCTGTCCCAATGTATGGGATGGCGGA...  \n",
       "193944  CGCACCAATAGGATCCTCCCGAATCAACCCTGACCCCTCTCCTTCA...  \n",
       "193945  TCCGATAAAATCACCTTCCACCCTTACTACACAATCAAAGACGCCC...  \n",
       "193946  TATGACCCTGAAGTAGGAACCAGATGTCGGATACAGTTCACTTTAG...  \n",
       "\n",
       "[193947 rows x 17 columns]"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "df8961ba-b227-49b1-80fd-9e3bb31b2c47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total unique gene IDs: 22114\n",
      "Total rows: 193947\n",
      "                  Name Description       PC1       PC2       PC3       PC4  \\\n",
      "17  ENSG00000237491.10   LINC01409 -0.128374 -0.012855 -0.002552 -0.009135   \n",
      "18  ENSG00000237491.10   LINC01409 -0.128374 -0.012855 -0.002552 -0.009135   \n",
      "19  ENSG00000237491.10   LINC01409 -0.128374 -0.012855 -0.002552 -0.009135   \n",
      "20  ENSG00000237491.10   LINC01409 -0.128374 -0.012855 -0.002552 -0.009135   \n",
      "21  ENSG00000237491.10   LINC01409 -0.128374 -0.012855 -0.002552 -0.009135   \n",
      "\n",
      "         PC5       PC6       PC7       PC8       PC9      PC10  \\\n",
      "17 -0.001463 -0.009729 -0.001872  0.001706  0.003546  0.000692   \n",
      "18 -0.001463 -0.009729 -0.001872  0.001706  0.003546  0.000692   \n",
      "19 -0.001463 -0.009729 -0.001872  0.001706  0.003546  0.000692   \n",
      "20 -0.001463 -0.009729 -0.001872  0.001706  0.003546  0.000692   \n",
      "21 -0.001463 -0.009729 -0.001872  0.001706  0.003546  0.000692   \n",
      "\n",
      "    cleaned_gene_id          gene_id  transcript_id_x  \\\n",
      "17  ENSG00000237491  ENSG00000237491  ENST00000457084   \n",
      "18  ENSG00000237491  ENSG00000237491  ENST00000655765   \n",
      "19  ENSG00000237491  ENSG00000237491  ENST00000412115   \n",
      "20  ENSG00000237491  ENSG00000237491  ENST00000670700   \n",
      "21  ENSG00000237491  ENSG00000237491  ENST00000429505   \n",
      "\n",
      "                        transcript_id_y  \\\n",
      "17  ENSG00000237491::1:778239-779239(+)   \n",
      "18  ENSG00000237491::1:778239-779239(+)   \n",
      "19  ENSG00000237491::1:778239-779239(+)   \n",
      "20  ENSG00000237491::1:778239-779239(+)   \n",
      "21  ENSG00000237491::1:778239-779239(+)   \n",
      "\n",
      "                                             sequence  \n",
      "17  TGCACACCTGGATAGTTTTTTAAATTTTTTTAGACAGGGTTTACCT...  \n",
      "18  TGCACACCTGGATAGTTTTTTAAATTTTTTTAGACAGGGTTTACCT...  \n",
      "19  TGCACACCTGGATAGTTTTTTAAATTTTTTTAGACAGGGTTTACCT...  \n",
      "20  TGCACACCTGGATAGTTTTTTAAATTTTTTTAGACAGGGTTTACCT...  \n",
      "21  TGCACACCTGGATAGTTTTTTAAATTTTTTTAGACAGGGTTTACCT...  \n"
     ]
    }
   ],
   "source": [
    "# Check how many unique gene IDs are duplicated\n",
    "print(f\"Total unique gene IDs: {final_merged_df['gene_id'].nunique()}\")\n",
    "print(f\"Total rows: {final_merged_df.shape[0]}\")\n",
    "\n",
    "# Display some duplicate examples\n",
    "duplicates = final_merged_df[final_merged_df['gene_id'].duplicated(keep=False)]\n",
    "print(duplicates.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "71949262-f53f-4966-9702-27208cc676b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape after removing duplicates: (22114, 17)\n"
     ]
    }
   ],
   "source": [
    "# Drop duplicates, keeping the first occurrence of each gene_id\n",
    "unique_final_df = final_merged_df.drop_duplicates(subset='gene_id', keep='first')\n",
    "\n",
    "# Verify the results\n",
    "print(f\"Shape after removing duplicates: {unique_final_df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "088948c3-b2b4-49d1-a73b-3314f6c3ce41",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_final_df = unique_final_df.drop(columns=[\"cleaned_gene_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "91167c7d-f73e-4cd6-971e-973309202a50",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_final_df = unique_final_df.drop(columns=[\"gene_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "b02ffd20-f84e-4eac-8d53-320d902ad6eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_final_df.to_csv(\"final_expression_with_sequences.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b3afd88-d562-4d9d-9fc7-34598bda5887",
   "metadata": {},
   "source": [
    "#### The section below represents commands run to process some of the data and to run BEDtools in conjunction with this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2be088aa-675f-4a3a-9057-62b657379544",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "Getting Genomic Sequences from Reference Genome (500bp upstream and downstream of each TSS):\n",
    "\n",
    "# Prepare a list of Ensembl IDs (one per line) from the GTEx data\n",
    "cut -f1 final_gene_expression_data.csv | sed 's/\\\\..*$//' > gtex_transcript_ids.txt\n",
    "\n",
    "# Extract Only the First Column, Remove the Header Line, and Remove Version Numbers\n",
    "cut -d',' -f1 gtex_transcript_ids.txt | tail -n +2 | sed 's/\\.[0-9]*$//' > gtex_transcript_ids_no_version.txt\n",
    "\n",
    "# Use awk to extract the gene_id and transcript_id pairs from the GTF file\n",
    "awk '$3 == \"transcript\" {match($0, /gene_id \"([^\"]+)\"/, g); match($0, /transcript_id \"([^\"]+)\"/, t); if (g[1] && t[1]) print g[1] \"\\t\" t[1]}' Homo_sapiens.GRCh38.113.gtf > gene_to_transcript_map.txt\n",
    "\n",
    "# Filter the gene_to_transcript_map.txt file to include only the gene IDs present in the gtex_transcript_ids_clean.txt file\n",
    "grep -Ff gtex_transcript_ids_no_version.txt gene_to_transcript_map.txt > filtered_gene_to_transcript_map.txt\n",
    "\n",
    "# Extract the transcript IDs\n",
    "cut -f2 filtered_gene_to_transcript_map.txt > transcript_ids.txt\n",
    "\n",
    "#Filter the GTF file:\n",
    "awk '$3 == \"transcript\"' Homo_sapiens.GRCh38.113.gtf | grep -Ff transcript_ids.txt > filtered.gtf\n",
    "\n",
    "# Convert the filtered GTF file to a BED file while extending regions by 500 base pairs upstream and downstream\n",
    "awk 'BEGIN {OFS=\"\\t\"} {match($0, /transcript_id \"([^\"]+)\"/, t); if (t[1]) print $1, $4-1, $5, t[1], \".\", $7}' filtered_genes.gtf > gtex_transcripts.bed\n",
    "\n",
    "# Use BEDTools to extract sequences from the reference genome\n",
    "bedtools getfasta -fi Homo_sapiens.GRCh38.dna.primary_assembly.fa -bed gtex_transcripts.bed -fo gtex_sequences.fa -name -s\n",
    "\n",
    "# Convert fasta file to csv format\n",
    "awk 'BEGIN {OFS=\",\"} /^>/ {if (seq) print id, seq; id=substr($1, 2); seq=\"\"} !/^>/ {seq=seq $0} END {print id, seq}' gtex_sequences.fa > gtex_sequences.csv\n",
    "\n",
    "\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
